회원가입, 로그인 기능 구현

tlqkrus012345 opened the pull request:
- 향후 다양한 로그인 방식을 지원할 수 있도록 로그인 인터페이스(LoginService)를 도입하고, 세션 기반 로그인 기능을 구현한 SessionLoginService 구현체를 작성했습니다.

- Spring Security가 제공하는 세션 기능 대신, 세션에 대한 이해를 높이기 위해 직접 세션 관리 로직을 수동으로 구현했습니다.

- 회원 엔티티는 DB 매핑에 집중하고, 입력 검증은 DTO 클래스에서 담당하도록 책임을 분리했습니다.

- 회원 스키마입니다
![스크린샷 2025-05-24 오후 5 13 48](https://github.com/user-attachments/assets/3a6c2d26-bca4-4a86-9c9b-be29730c4c49)

tlqkrus012345 committed:
build: 회원가입 기능을 위한 라이브러리 추가

- JPA: 객체 중심 개발을 위해 추가
- H2: 초기 테스트용 데이터베이스를 위해 추가
- Security: 비밀번호 암호화 기능 구현을 위해 추가
- Validation: 입력값 유효성 검사를 위해 추가

Resolvers: #3

tlqkrus012345 committed:
chore: H2 DB 및 JPA 설정 추가

- H2 DB TCP 연결 설정
- JPA SQL 로그 출력 설정
- 스키마 자동 생성

Resolves: #3

tlqkrus012345 committed:
build: 롬북 라이브러리 추가

Resolves: #3

tlqkrus012345 committed:
chore: h2 DB 인메모리로 전환 설정  


Resolves: #3

tlqkrus012345 committed:
feat: 회원 엔티티 구현

- 회원 정보 컬럼 및 생성자 구현
- 성별, 회원역할 Enum 클래스 구현

Resolves: #3

tlqkrus012345 committed:
feat: 회원가입 기능 구현

- 비밀번호 BCrypt로 암호화하여 저장 기능 구현
- 회원가입 엔드포인트 구현
- 회원가입 request, response dto 구현

Resolves: #3

tlqkrus012345 committed:
refactor: 회원 정적 팩토리 메소드 수정

- 비밀번호 암호화를 위해 생성자 매개변수 및  내부 수정

Resolves: #3

tlqkrus012345 committed:
feat: Security 설정 파일 구현

- H2 DB 접근 허용
- CSRF 비활성화

Resolves: #3

tlqkrus012345 committed:
test: 회원 가입 및 입력값 검증 테스트

- DB에 insert 됨을 검증하기 위해 join() 메서드는 통합 테스트로 구현
- 입력값 테스트는 공백, 패턴을 테스트하였고 단위 테스트로 구현

Resolves: #3

tlqkrus012345 committed:
feat: 생성자 구현

- 테스트를 위해 생성자 구현

Resolves: #3

tlqkrus012345 committed:
feat: 빈생성자 룸북 어노테이션 추가

- dto 직렬화에 필요한 빈 생성자를 위해 추가

Resolves: #3

tlqkrus012345 committed:
feat: 세션 기반 로그인 기능 구현

- 다른 로그인 방식으로 전환을 위해 인터페이스를 생성 및 수동 빈 등록 설정 클래스 구현
- Security를 사용하지 않고 수동 Session 기반으로 구현
- 로그인에 필요한 dto 클래스 구현
- 로그인 엔드포인트 구현

Resolves: #3

tlqkrus012345 committed:
test: 세션 기반 로그인 테스트

- 로그인 성공, 로그인 실패, 로그 아웃 성공 테스트
- 로그인 엔드포인트, 회원 가입 엔드포인트 테스트 @WebMvcTest로 경량화 구현
- 테스트 환경을 위한 Security Config 클래스 구현

Resolves: #3

tlqkrus012345 committed:
test: findByEmail 테스트

Resolves: #3

tlqkrus012345 committed:
test: 회원 컨트롤러 API 테스트

- 회원가입, 로그인, 로그아웃

Resolves: #3

f-lab-martin commented on paytogether/src/main/java/com/paytogether/config/LoginConfig.java:
```diff
@@ -0,0 +1,21 @@
+@Configuration
+public class LoginConfig {
+
+  @Bean
+  public LoginService loginService(
+      MemberRepository memberRepository,
+      PasswordEncoder passwordEncoder,
+      HttpSession httpSession) {
+    return new SessionLoginService(memberRepository, passwordEncoder, httpSession);
+  }
```
SessionLoginService 클래스에서 빈 생성을 명시해도 되는데 여기서 하신 이유가 있으실까요?

f-lab-martin commented on paytogether/src/main/java/com/paytogether/domain/member/entity/Gender.java:
```diff
@@ -0,0 +1,15 @@
+  FEMALE("여"),
+  OTHER("그외");
+
+  private final String text;
```
text보다 좀더 의미있는 표현으로 바꿔보면 어떨까요?

f-lab-martin commented on paytogether/src/main/java/com/paytogether/domain/member/entity/Gender.java:
```diff
@@ -0,0 +1,15 @@
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+
+@Getter
```
해당 enum에 Getter가 필요한 이유가 있을까요?

f-lab-martin commented on paytogether/src/main/java/com/paytogether/domain/member/entity/Member.java:
```diff
@@ -0,0 +1,71 @@
+    this.role = role;
+  }
+
+  public static Member createMember(MemberJoinRequest request, String encodedPassword) {
+    return Member.builder()
```
Controller에서 사용되는 DTO가 domain 클래스까지 그대로 넘어오고 있습니다. 필요한 필드를 파라미터로 받도록 수정해보면 어떨까요?

f-lab-martin commented on paytogether/src/main/java/com/paytogether/domain/member/entity/Member.java:
```diff
@@ -0,0 +1,71 @@
+
+  private boolean isActive = true;
+
+  @Builder
```
빌터패턴을 직접 구현할 때와 lombok을 사용했을 때의 차이가 뭘까요?

f-lab-martin commented on paytogether/src/main/java/com/paytogether/domain/member/entity/Member.java:
```diff
@@ -0,0 +1,71 @@
+  @Enumerated(EnumType.STRING)
+  private MemberRole role;
+
+  private boolean isActive = true;
```
isActive 는 어떤 의미의 컬럼인가요? active된 시점이나 deactivate된 시점은 없어도 될까요? 
가입한 시기나 탈퇴한 시기를 관리할 필요는 없을까요?

f-lab-martin commented on paytogether/src/main/java/com/paytogether/domain/member/service/LoginService.java:
```diff
@@ -0,0 +1,12 @@
+
+import jakarta.servlet.http.HttpSession;
+
+public interface LoginService {
+
+  void login(String email, String password);
+
+  void logout();
+
+  boolean sessionExists(HttpSession session);
+}
```
LoginService 를 인터페이스로 분리하신 이유가 있으실까요?

f-lab-martin commented on paytogether/src/main/java/com/paytogether/domain/member/entity/Member.java:
```diff
@@ -0,0 +1,71 @@
+@Getter
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@Entity
+public class Member {
```
Member에 대한 validation은 없어도 괜찮을까요?

f-lab-martin commented on paytogether/src/main/java/com/paytogether/domain/member/service/SessionLoginService.java:
```diff
@@ -0,0 +1,50 @@
+
+public class SessionLoginService implements LoginService {
+
+  private final MemberRepository memberRepository;
+  private final PasswordEncoder passwordEncoder;
+  private final HttpSession httpSession;
+  public static final String LOGIN_MEMBER = "LOGIN_MEMBER";
+
```
상수는 클래스 내에서 최상단에 선언해주세요~! 
선언하는 순서들에 대해 자바 가이드를 찾아보셔도 좋습니다.
전에 공유해드린  google java style 가이드를 설정해주시면 좋을 것 같습니다.

f-lab-martin commented on paytogether/src/main/java/com/paytogether/domain/member/service/SessionLoginService.java:
```diff
@@ -0,0 +1,50 @@
+
+  @Override
+  public void login(String email, String rawPassword) {
+    Optional<Member> optionalMember = memberRepository.findByEmail(email);
+    if (optionalMember.isEmpty()) {
+      throw new IllegalArgumentException();
+    }
+    Member presentMember = optionalMember.get();
```
이부분 좀더 간결하게 리팩토링해보면 어떨까요?
IllegalArgumentException을 그대로 사용하기보다는 Custom한 예외를 발생시키도록 수정해보면 어떨까요?

f-lab-martin commented on paytogether/src/main/java/com/paytogether/domain/member/service/SessionLoginService.java:
```diff
@@ -0,0 +1,50 @@
+
+  @Override
+  public boolean sessionExists(HttpSession session) {
+    Object loginMemberId = session.getAttribute(LOGIN_MEMBER);
+    if (loginMemberId != null) {
+      return true;
+    }
+    return false;
```
nullable한 값이라면 Optional.ofNullable을 사용해보면 어떨까요?

f-lab-martin commented on paytogether/src/main/java/com/paytogether/domain/member/service/SessionLoginService.java:
```diff
@@ -0,0 +1,50 @@
+  }
+
+  @Override
+  public void login(String email, String rawPassword) {
```
여러명의 사용자가 로그인했을 경우에는 어떻게 처리하는게 좋을까요?

f-lab-martin commented on paytogether/src/main/java/com/paytogether/domain/member/service/SessionLoginService.java:
```diff
@@ -0,0 +1,50 @@
+    httpSession.setAttribute(LOGIN_MEMBER, presentMember.getId());
+  }
+
+  @Override
+  public void logout() {
+    httpSession.invalidate();
+  }
+
```
여러명의 사용자가 로그아웃하려면 어떻게 해야할까요?

f-lab-martin commented on paytogether/src/test/java/com/paytogether/api/MemberControllerTest.java:
```diff
@@ -0,0 +1,133 @@
+
+
+  @Test
+  void login_success() throws Exception {
+    MemberLoginRequest request = new MemberLoginRequest("email@spring.com", "!!password");
+
+    mockMvc.perform(post("/member/login")
+            .contentType(MediaType.APPLICATION_JSON)
+            .content(objectMapper.writeValueAsString(request)))
+        .andDo(print())
+        .andExpect(status().isOk());
```
로그인 성공 테스트로, HttpStatus만 200인 것만 확인해도 충분할까요?

f-lab-martin commented on paytogether/src/test/java/com/paytogether/api/MemberControllerTest.java:
```diff
@@ -0,0 +1,133 @@
+        .andExpect(status().isOk());
+  }
+
+  private Member createMember(String email, String password) {
+    MemberJoinRequest request = MemberJoinRequest.builder()
+        .age(100)
+        .name("pay")
+        .gender(Gender.MALE)
+        .address("seoul")
+        .password(password)
+        .email(email)
+        .phoneNumber("01012345678")
+        .build();
+    return Member.createMember(request, password);
+  }
```
사용되지 않는 메소드는 삭제해주세요~

f-lab-martin commented on paytogether/src/test/java/com/paytogether/domain/member/repository/MemberRepositoryTest.java:
```diff
@@ -0,0 +1,37 @@
+
+    Optional<Member> result = memberRepository.findByEmail("pay@spring.com");
+
+    assertThat(result).isNotNull();
```
테스트코드 작성시에는 given,when,then pattern에 맞춰서 작성해보면 어떨까요?
result가 단순히 null이 아닌 것만 테스트해도 괜찮을까요?

f-lab-martin commented on paytogether/src/test/java/com/paytogether/domain/member/service/MemberJoinRequestValidationTest.java:
```diff
@@ -0,0 +1,97 @@
+
+    Set<ConstraintViolation<MemberJoinRequest>> violations = validator.validate(request);
+
+    assertThat(violations).hasSize(1);
+    assertThat(violations)
+        .extracting("message")
+        .contains("공백일 수 없습니다");
+  }
```
하나의 테스트케이스에는 하나의 assert구문을 쓰도록 테스트를 작게 나누시는 것이 좋습니다. 해당 구문을 한번에 동작하게 하려면 어떻게 하면 좋을지 찾아보시면 좋을 것 같습니다. 한 가지 방법으로는 assertAll이 있지만 다른 방법도 있는지 더 찾아보시면 좋을 것 같습니다.

f-lab-martin commented on paytogether/src/test/java/com/paytogether/domain/member/service/SessionLoginServiceTest.java:
```diff
@@ -0,0 +1,100 @@
+
+    loginService.login(email, password);
+
+    Long afterLoginMemberId = (Long) httpSession.getAttribute(LOGIN_MEMBER);
```
LOGIN_MEMBER를 테스트코드에서도 알지 않아도 테스트가 가능하도록 리팩토링해보면 어떨까요?

f-lab-martin reviewed (commented):
멘토링 시간에 말씀드렸던 것과 함께 코멘트 남깁니다.

tlqkrus012345 committed:
refactor: enum 클래스 수정

text -> explain 으로 영문 내용을 설명한다는 의미로 변경
사용하지 않는 @Getter 어노테이션 제거

Resolves: #3

tlqkrus012345 committed:
remove: LoginConfig 클래스 삭제

- 추후 다양한 로그인 전략으로 교체할 수 있도록 LoginService 구현체를 수동으로 등록했으나, 현재는 단일 구현체만 존재하고 교체 계획도 없어 컴포넌트 스캔을 통한 자동 빈 등록 방식으로 변경

Resolves: #3

tlqkrus012345 committed:
feat: Security Config 세션 설정

- Security의 자동 세션 생성 기능을 사용하지 않기 위해 SessionCreationPolicy.NEVER 로 설정
따라서 SessionLoginService를 통해 세션이 생성되도록 설정

Resolves: #3

tlqkrus012345 committed:
refactor: 회원 생성 정적 패토리 메서드

- 회원 생성 정적 팩토리 메서드 dto가 아닌 필요한 매개변수로 생성하게 변경
- 테스트 코드에 적용 가능한 생성 메서드 추가
- 회원 서비스 필요한 매개변수를 전달할 수 있도록 변경

Resolves: #3

tlqkrus012345 committed:
refactor: sessionExists 메서드

- null 체크에 대한 가독성을 위해 Optional.ofNullable로 변경

Resolves: #3

tlqkrus012345 committed:
refactor: 로그인, 로그아웃 기능

- 기존에 로그인한 상태이고, 다시 로그인을 시도하면 기존 세션은 삭제하고 신규 발급

Resolves: #3

tlqkrus012345 committed:
build: redis 의존성

- redis에 저장하기 위해 spring-session-data-redis 추가
- spring-session을 위해 spring-session-data-reids 추가

Resolves: #3

tlqkrus012345 committed:
chore: redis 설정

- redis 연결 설정
- redis session namespace 설정

Resolves: #3

tlqkrus012345 committed:
test: SoftAssertions 적용

- 테스트 가독성을 위해 적용

Resolves: #3

tlqkrus012345 committed:
test: assert구문 변경

- assert구문 테스트 목적에 맞게 변경

Resolves: #3

tlqkrus012345 committed:
chore: 패키지 구조 변경

- api 패키지 domain 패키지 하위로 변경
- 공통 패키지 생성 후 TestController 이동

Resolves: #3

tlqkrus012345 committed:
feat: 공통 필드를 위한 BaseEntity 클래스

- 생성날짜, 수정날짜 관리를 위해 적용

Resolves: #4

tlqkrus012345 committed:
test: JpaAuditing

- 생성날짜와 수정날짜 삽입 확인을 위해 구현

Resolves: #4

tlqkrus012345 committed:
feat: 전반적으로 사용하는 공통 모듈 기능

- 공통 응답 DTO
- 글로벌 예외 핸들러
- 애플리케이션 커스텀 에러
- 에러 응답 DTO

Resolves: #4

tlqkrus012345 committed:
refactor: 로그인 응답 반환

- 테스트, Api 응답 의미 있는 반환을 위해 로그인 리턴 타입 void -> LoginResponse 로 변경

Resolves: #3

tlqkrus012345 committed:
test: 회원 컨트롤러 테스트 수정

- login 테스트 반환 값 확인 테스트

Resolves: #3

tlqkrus012345 committed:
chore: api 테스트 패키지 변경

Resolves: #3

tlqkrus012345 committed:
test: 테스트 전용 상수 클래스 구현

- 테스트와 서비스 상수 분리하기 위해 테스트 전용 상수 클래스 구현

Resolves: #3

tlqkrus012345 commented:
1) LoginService 인터페이스를 정의하고 SessionLoginService 구현체를 생성한 이유
추후에 세션 기반의 로그인에서 JWT를 활용한 로그인 서비스를 적용할 수 있는 가능성 때문에 인터페이스를 정의하고 수동으로 빈을 등록하는 방식을 택했습니다.

2) 사용하지 않는 메소드, 의미를 파악하기 어려운 변수
의미있는 변수 적용 text -> explain , 사용하지 않는 메소드, 어노테이션 제거

3) Controller에서 사용하는 DTO가 Domain 침투 
서비스 레이어에서 필요한 필드만 추출하여 전달하는 방식으로 변경

4) 롬북 빌더, 사용자 정의 빌더 차이
두 빌더의 차이는 룸북 빌더를 통해 생성하면 전달되는 값에 대한 검증이 안된다는 점이 있습니다. 반면에 사용자 정의 빌더는 내부에서 검증 관련 기능을 추가하여 검증할 수 있습니다. 사용자 생성에 있어서 더 견고한 검증이 필요한 빌더 패턴이 필요하다면 사용자 정의 빌더를 사용할 것이고, 개발의 편의성을 추구한다면 롬북 빌더를 사용할 수 있습니다.

5) Member Validation
저는 컨트롤러에서 사용하는 DTO에서 검증하고 있습니다. 엔티티 클래스는 DB 컬럼 매핑과 비즈니스 로직 중심의 역할을 담당하므로,
검증 로직을 중복해서 두기보다는 입력 전 단계에서 걸러주는 방식이 추후에 요구사항 변경 시 더 편리하다고 판단했습니다.다만, 검증이 중요한 필드나 데이터 무결성이 중요한 컬럼에 대해서는 필요시 @Column(nullable = false) 같은 DB 레벨 제약 조건을 추가도 고려하고 있습니다.

6) google java style 적용

7) custom 예외 적용
더 의미있는 에러를 반환하기 위해 커스텀한 에러 클래스를 구현하고 이를 핸들링하는 클래스를 구현했습니다.

8) 로그인 반환 타입
void -> LoginResponse 로 변경하여 테스트, 클라이언트 확인에 있어서 더 유용하게 사용할 수 있도록 변경하였습니다.

9) 테스트 코드 
give, when, then 패턴 적용
assert 구문 최소화를 위해 테스트 목적에 맞는 assert 구문으로 변경 및 작은 assert 구문으로 나누는게 힘들다고 판단된 테스트 코드는SoftAssertion 적용하였습니다.

10) LOGIN_MEMBER 서비스 상수
서비스에서 사용되는 상수가 테스트 코드에서도 사용되어 테스트용 상수 모음 클래스를 정의했습니다.

tlqkrus012345 committed:
test: 테스트 예외 수정

Resolves: #3

tlqkrus012345 committed:
feat: validation exception

- validation exception 핸들링 하는 메서드 구현

Resolves: #3

tlqkrus012345 committed:
build: testcontainer

- 테스트 환경의 일관성을 위해 testcontainer 추가

Resolves: #3

tlqkrus012345 committed:
feat: redis container 설정

- 테스트 환경의 일관성을 위해 redis test container 설정 추가

Resolves: #3

tlqkrus012345 committed:
test: 예외 수정

Resolves: #3

tlqkrus012345 committed:
chore: 패키지 구조 변경

- domain 패키지 제거
- 루트 디렉토리 변경
- exception 패키지 변경

Resolves: #3

tlqkrus012345 committed:
remove: BaseEntity 클래스, 설정

- common 모듈에 대한 중복 제거 가능한 클래스 제거

Resolves: #3

tlqkrus012345 committed:
refactor: 회원 생성, 비활성 컬럼

- jpaauditing 제거하고 컬럼을 명시적으로 입력하게 리팩토링
- 생서날짜, 비활성날짜 테스트 일관성을 위해 컨트롤러에서 전달

Resolves: #3

Resolves: #3

tlqkrus012345 committed:
test: 회원 생성 코드 수정

- 회원 생성 시 날짜 전달하게 수정

Resolves: #3

tlqkrus012345 committed:
feat: 회원 엔티티 DB 제약 조건

- 회원 엔티티에서 DB 제약 조건 책임 부여

Resolves: #3

tlqkrus012345 committed:
test: 회원 엔티티 DB 제약 조건 추가에 의해 수정

Resolves: #3

tlqkrus012345 committed:
refactor: 계층간 dto 분리

- 레이어별 의존관계를 고려하여 dto 분리 service(command, result) controller(request, response)

Resolves: #3

tlqkrus012345 committed:
rename: 파일 이름 오타 변경

Resolves: #3

tlqkrus012345 committed:
test: 계층별 dto 변경으로 인한 수정

Resolves: #3

tlqkrus012345 committed:
fix: 상태 코드 반환 오류

- 생성자 수정

Resolves: #3

tlqkrus012345 committed:
rename: 디렉토리 이름

- api -> controller 변경

Resolves: #3

tlqkrus012345 committed:
chore: 패키지 구조 변경
